src/main/java/cn/nukkit/event/entity/EntityDamageEvent.java

```diff
@@ -109,6 +116,56 @@ public void setAttackCooldown(int attackCooldown) {
         this.attackCooldown = attackCooldown;
     }
 
+    @PowerNukkitOnly
+    @Since("FUTURE")
+    @Nonnull
+    public SideEffect[] getSideEffects() {
+        SideEffect[] sideEffects = this.sideEffects;
+        if (sideEffects.length == 0) {
+            return sideEffects;
+        }
+        return Arrays.stream(sideEffects)
+                .map(SideEffect::clone)
+                .toArray(SideEffect[]::new)
+        ;
+    }
+
+    @PowerNukkitOnly
+    @Since("FUTURE")
+    public void setSideEffects(@Nonnull SideEffect... sideEffects) {
+        this.sideEffects = Arrays.stream(sideEffects)
+                .filter(Objects::nonNull)
+                .map(SideEffect::clone)
+                .toArray(SideEffect[]::new)
+        ;
+    }
+
+    @PowerNukkitOnly
+    @Since("FUTURE")
+    public void setSideEffects(@Nonnull Collection<SideEffect> sideEffects) {
+        this.setSideEffects(sideEffects.toArray(SideEffect.EMPTY_ARRAY));
+    }
+
+    @PowerNukkitOnly
+    @Since("FUTURE")
+    public void addSideEffects(@Nonnull SideEffect... sideEffects) {
+        this.sideEffects = Stream
```

Comment:
Isn't this just a very verbose `Streams.concat` ?
like, `flatmapping` into `id` is just concatenation.
you're taking an `[[a]]` and converting it into an `[a]`.

```
:t concat
concat :: Foldable t => t [a] -> [a]
```

thats already implemented in the java streams api under the pure static function `Streams.concat`.

the code will end up looking like this:
```java
Stream<SideEffect> safeInput = Arrays.stream(sideEffects) // tail
                .filter(Objects::nonNull)
                .map(SideEffect::clone);

this.sideEffects = Stream.concat(Arrays.stream(this.sideEffects), safeInput).toArray(SideEffect[]::new);
```


i've also noted the constant usage of the `map clone $ filter (not . null) array` snippet.
we should make that into an helper function (perhaps `functionalCopy`) , but thats left to another day. code will be even more clean if this suggestion is also applied.

```java
this.sideEffects = Stream.concat(Arrays.stream(this.sideEffects), FunctionalUtils.functionalCopy(sideEffects)).toArray(SideEffect[]::new);

Reply:
gh decided to mark the wrong lines for me. the code i'm referring to is this snippet here
https://github.com/PowerNukkit/PowerNukkit/blob/d67b5c9a36103843c502d0630f1a0eb76af93483/src/main/java/cn/nukkit/event/entity/EntityDamageEvent.java#L152-L160